aspnetcore/fundamentals/host/generic-host.md

```diff
@@ -1,31 +1,494 @@
+* [IHostEnvironment / IWebHostEnvironment](#ihostenvironment)
+
+For a list of all framework-provided services, see <xref:fundamentals/dependency-injection#framework-provided-services>.
+
+## IHostApplicationLifetime
+
+Inject the <xref:Microsoft.Extensions.Hosting.IHostApplicationLifetime> (formerly `IApplicationLifetime`) service into any class to handle post-startup and graceful shutdown tasks. Three properties on the interface are cancellation tokens used to register app start and app stop event handler methods. The interface also includes a `StopApplication` method.
+
+The following example is an `IHostedService` implementation that registers the `IApplicationLifetime` events:
+
+[!code-csharp[](generic-host/samples-snapshot/3.x/LifetimeEventsHostedService.cs?name=snippet_LifetimeEvents)]
+
+## IHostLifetime
+
+The <xref:Microsoft.Extensions.Hosting.IHostLifetime> implementation controls when the host starts and when it stops. The last implementation registered is used.
+
+<xref:Microsoft.Extensions.Hosting.Internal.ConsoleLifetime> is the default `IHostLifetime` implementation. `ConsoleLifetime`:
+
+* listens for Ctrl+C/SIGINT or SIGTERM and calls <xref:Microsoft.Extensions.Hosting.IApplicationLifetime.StopApplication*> to start the shutdown process.
+* Unblocks extensions such as [RunAsync](#runasync) and [WaitForShutdownAsync](#waitforshutdownasync).
+
+## IHostEnvironment
+
+Inject the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> service into a class to get information about the following:
+
+* [ApplicationName](#applicationname)
+* [EnvironmentName](#environmentname)
+* [ContentRootPath](#contentrootpath)
+
+Web apps implement the `IWebHostEnvironment` interface, which inherits `IHostEnvironment` and adds:
+
+* [WebRootPath](#webroot)
+
+## Host configuration
+
+Host configuration is used for the following:
+
+* The properties of the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> implementation.
+* The [URLs](#urls) setting in web apps. The host passes this value to the servers through an `IServerAddresses` API.
+
+Host configuration automatically flows to app configuration ([ConfigureAppConfiguration](#configureappconfiguration).
+
+To add host configuration, call <xref:Microsoft.Extensions.Hosting.HostBuilder.ConfigureHostConfiguration*> on `IHostBuilder`. `ConfigureHostConfiguration` can be called multiple times with additive results. The host uses whichever option sets a value last on a given key.
+
+The environment variable provider with prefix `DOTNET_` is included by default. For web apps, the environment variable provider with prefix `ASPNETCORE_` is added. The prefix is removed when the environment variables are read. For example, the environment variable value for `ASPNETCORE_ENVIRONMENT` becomes the host configuration value for the `environment` key.
+
+The following example creates host configuration:
+
+[!code-csharp[](generic-host/samples-snapshot/3.x/Program.cs?name=snippet_HostConfig)]
+
+> [!TIP]
```

Comment:
Too much bling for me. This also seems TMI for an advanced topic. Perhaps something like my next proposal.